openkruise/kruise
# Auto resize pvc when asts volumeClaimTemplates expand

### Ⅰ. Describe what this PR does
imp the proposal https://github.com/openkruise/kruise/pull/1651
can preview in https://github.com/ABNER-1/kruise/blob/add-proposal-resize-asts-pvc/docs/proposals/20240626-asts-volume-resize-zh.md
api changes:
1. add `VolumeClaimUpdateStrategy` in `StatefulSetSpec`
2. add `VolumeClaimTemplates` in `StatefulSetStatus`

core changes
1. Monitor the PVC update event and extract the StatefulSet name from the annotation. This annotation will be included during the PVC patching process.
3. In the core function `rollingUpdateStatefulSetPods`, if the feature gate is enabled, treat pods with an unready PVC as unavailable. This action may disrupt the update process for additional pods.
4. Before attempting an update, if the feature gate is enabled, verify whether the PVC requires an update and apply the patch if the storage class supports volume expansion.
5. After the PVC is patched, check if the PVC is ready (i.e., if the capacity in the status is greater than or equal to the specification) or if the PVC is in the `FileSystemResizePending` condition, as described in the [Kubernetes blog post](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/).
6. Update the PVC status during the status update process to provide users with feedback on the outcome of the PVC resizing operation.

### Ⅱ. Does this pull request fix one issue?

fixes #905 

### Ⅲ. Describe how to verify it


### Ⅳ. Special notes for reviews
